src/main/scala/SchedulerImp.scala

```diff
@@ -0,0 +1,38 @@
+import java.util.concurrent.PriorityBlockingQueue
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SchedulerImp(
+    val minHeap: PriorityBlockingQueue[ScheduledTask]
+) extends Scheduler {
+
+  @volatile var running: Boolean = false
+
+  override def schedule(task: () => Unit, timestamp: Long): Unit = {
+    minHeap.offer(ScheduledTask(task, timestamp))
+  }
+
+  override def start()(implicit ec: ExecutionContext): Unit = {
+    if (!running) {
+      running = true
+      Future {
+        while (running) {
+          // put and take are blocking
+          val nextTask = minHeap.take()
+          if (nextTask.timestamp < System.currentTimeMillis) {
+            Future {
```

Comment:
dispatch the task execution to a worker thread in the thread pool to not block the task-scheduler.